Sources/ApodiniNetworking/URI.swift

```diff
@@ -0,0 +1,123 @@
+//
+// This source file is part of the Apodini open source project
+//
+// SPDX-FileCopyrightText: 2019-2021 Paul Schmiedmayer and the Apodini project authors (see CONTRIBUTORS.md) <paul.schmiedmayer@tum.de>
+//
+// SPDX-License-Identifier: MIT
+//
+
+import Foundation
+import ApodiniUtils
+
+
+/// A URL
+public struct URI: LosslessStringConvertible {
+    public enum Scheme: String {
+        case http
+        case https
+    }
+    public let scheme: Scheme
+    public let hostname: String
+    public let port: Int?
+    public let path: String
+    public let rawQuery: String
+    public let fragment: String
+    public let queryItems: [String: String?]
+    
+    
+    public init(
+        scheme: Scheme,
+        hostname: String,
+        port: Int? = nil,
+        path: String = "/",
+        rawQuery: String = "",
+        fragment: String = ""
+    ) {
+        self.scheme = scheme
+        self.hostname = hostname
+        self.port = port
+        self.path = path
+        self.rawQuery = rawQuery.hasPrefix("?") ? String(rawQuery.dropFirst()) : rawQuery
+        self.fragment = fragment
+        self.queryItems = .init(
+            uniqueKeysWithValues: (URLComponents(string: "?\(self.rawQuery)")?.queryItems ?? []).map { ($0.name, $0.value) }
```

Comment:
Am I correct that the [Precent-Encoding](https://en.wikipedia.org/wiki/Percent-encoding) (for encoding and decoding) through Foundations API, right?

Reply:
That was my assumption, yes. I would've expected the tests to fail if it wasn't.